### PR TITLE
Add support for container image pulling and unpacking

### LINKED ISSUE #12: Support container image pulling and unpacking
In order for `run0` to be self contained and pull container images from e.g. within a virtual machine, we'd like it to be able to create container bundles from a container URL

`run0 pull <container_url>`
`run0 mount <container_url> <mount_point>`

Ideally, this capability should be provided by a separate binary/tool, but that could be done as a further step.

See https://github.com/confidential-containers/image-rs

### PR DESCRIPTION
**Pull request by thomasgouveia**
This PR aims to add to `kaps` the ability to pull container images and unpack them into ready-to-use OCI bundles.

- [x] Pull image 
- [x] Store state about pulled images & layers
- [x] Create a bundle from an image ID
- [x] Handle  `pull` command options
- [x] Handle  `mount` command options

Closes #12

### CODE REVIEW

**Comment by thomasgouveia** [comment]
TODO 
- [x] Pull image 
- [ ] Store state about pulled images & layers
- [ ] Create a bundle from an image ID
- [ ] Handle commands options (`pull`, `bundle`)

**Comment by thomasgouveia** [comment]
Hello @sameo, finally the first version of my PR is ready to review! Sorry for the late, I was working on another project...

First, I'm currently dependent on Malo's PR (#16), so we have to merge it first before merging my PR, so I will adapt my dependencies in consequence.

My PR brings up the following features : 

- Pull an image from a public registry (private registries can be added in another PR easily I think)
- Create a bundle from an image

When this PR will be merged, I will do another PR to add the ability to list images, such as `kaps images list` or `kaps images`.

I'm waiting for your review and comments about my work!

File: oci-image/src/snapshots/mod.rs
Comment (sameo) [inline oci-image/src/snapshots/mod.rs]: Overlayfs is only one way of creating a container snapshot, and should be an implementation of a trait:

```rust
trait Snapshotter {
   pub fn mount(&mut self, layers: Vec<String>, mount_path: &Path) -> Result<MountPoint>;
   pub fn umount(&self, mount_point: &MountPoint) -> Result<()>;
}
```

Then your `ImageManager`would be generic and take a `S where S:Snapshotter`

File: oci-image/src/lib.rs
Comment (sameo) [inline oci-image/src/lib.rs]: With `snapshot`being generic, it is up to the caller to decide which snapshotting mechanism it wants to use. In other words, `new`here should take a `Snapshotter`instance as an argument and own it.

File: oci-image/src/lib.rs
Comment (sameo) [inline oci-image/src/lib.rs]: This is:

```rust
 let image = self
            .state
            .lock()
	    .await
            .get_image(image_id)
            .ok_or(Error::ImageNotFound(format!("foo")))?
            .clone();
```

File: oci-image/src/state.rs
Comment (sameo) [inline oci-image/src/state.rs]: Just use `image()`for an image getter.

File: oci-image/src/state.rs
Comment (sameo) [inline oci-image/src/state.rs]: ditto

File: oci-image/src/state.rs
Comment (sameo) [inline oci-image/src/state.rs]: And same here.

File: oci-image/src/state.rs
Comment (sameo) [inline oci-image/src/state.rs]: If I understand correctly what you're doing with the `snapshot` hash map is keeping track of an atomically incrementing index to use for the overlayfs `workdir`. Is that correct?

If it is, I don't think this should be handled this way. What you need is a snapshot index, and nothing else. So you could have the `State`carry an `AtomicUsize` and a `State` `snapshot_index`that `fetch_add`that index. Then when you create your `Snapshotter`you give it a `Arc::clone()`of your `State`and the snapshotter can then call `snapshot_index()` on it when it needs to create a new snapshot or mount.

File: oci-image/src/lib.rs
Comment (sameo) [inline oci-image/src/lib.rs]: `mount`should return a `Result<PathBuf>`.

Moreover, unless I'm missing something, we should not pass a `destination` parameter to this method. We ask it to mount an image, and it returns where it gets mounted. It is the caller's responsibility to then bind mount it somewhere else.

File: oci-image/src/state.rs
Comment (sameo) [inline oci-image/src/state.rs]: Do we really need that one?

File: oci-image/src/state.rs
Comment (sameo) [inline oci-image/src/state.rs]: As I dont think you need `state_file`at all, you could just call `map_or`on `serde_json::from_reader::<File, State>(file)`.

File: oci-image/src/state.rs
Comment (sameo) [inline oci-image/src/state.rs]: See https://doc.rust-lang.org/std/fs/struct.OpenOptions.html#method.create

File: oci-image/src/pull.rs
Comment (sameo) [inline oci-image/src/pull.rs]: I dont think this comment is accurate. This method returns a vector of layer metadata, which is a good thing.

File: oci-image/src/pull.rs
Comment (sameo) [inline oci-image/src/pull.rs]: This could be moved before the `async move` and avoid a whole `state` clone when the layer already exists.

**Review by sameo — COMMENTED** [review]
Thanks a lot for the PR @thomasgouveia, and sorry for the review delay.

This is a very solid piece of work already, but I added a few comments.

Another more generic comment that I'd like you to address, is to split the `kaps: add crate to manage OCI images` commit into smaller pieces:


1. The first one would bring the State module
2. The second one would bring the snapshots module
3. The third one would contain the puller
4. The last one would introduce the ImageManager

This would make it for a nicer split of work, and an easier one to review.

File: oci-image/src/pull.rs
Comment (thomasgouveia) [reply oci-image/src/pull.rs]: I think it's not possible because `state` needs to be locked. As I use the `tokio::sync::Mutex`, I have to use await and in consequence be in an async block?

Comment (sameo) [reply oci-image/src/pull.rs]: ah of course...tokio. Fine with me then.

File: oci-image/src/lib.rs
Comment (thomasgouveia) [reply oci-image/src/lib.rs]: I think I don't understand what we want for this method/command. In my mind, If I wish to run a container with `kaps`, I will first pull the image : 

```
$ kaps pull docker.io/amd64/alpine --name alpine
``` 

Then, I have to create a bundle based on my image, so : 

```
$ kaps mount alpine /tmp/alpine
```
The path `/tmp/alpine` here is where we want to mount the container bundle. But if we simply use: `kaps mount alpine`, the bundle will be mounted on a generated path inside the kaps working directory (`/var/lib/kaps`).

Then, I can simply run my container with : 

```
$ kaps run --bundle /tmp/alpine # or the path returned by the mount if the user hasn't specified a specific path
```

Is this ok?

File: oci-image/src/state.rs
Comment (thomasgouveia) [reply oci-image/src/state.rs]: Yes, because I need it to save the state. Do you think that we can handle this in a better way?

File: oci-image/src/lib.rs
Comment (sameo) [reply oci-image/src/lib.rs]: It's pretty close to what I have in mind. My idea was simpler:

```
$ kaps pull docker.io/amd64/alpine # no --name
```

```
# Mount just returns the path to where kaps mounted it.
# If the caller needs it to be somewhere else, it can bind mount that path.
$ kaps mount docker.io/amd64/alpine
/var/lib/kaps/snapshot/<id>/
```

Comment (thomasgouveia) [reply oci-image/src/lib.rs]: Ok, I'll remove the `destination` parameter. 

The `--name` can be used to give an identifier for the image. Shall we remove it also?

Comment (sameo) [reply oci-image/src/lib.rs]: > The `--name` can be used to give an identifier for the image. Shall we remove it also?

As long as things work without the name, I'm fine. Note that adding support for `--name` adds code complexity that you could avoid for now.

Comment (thomasgouveia) [reply oci-image/src/lib.rs]: Without the name, the id of the image is computed by hashing the reference. By using that method, I'm sure that my identifier is unique. So yes, without name, it works perfectly also.

File: oci-image/src/state.rs
Comment (sameo) [reply oci-image/src/state.rs]: If you're going to save it somewhere under a path you can reconstruct from your image id and a fixed prefix, why bother?

Comment (thomasgouveia) [reply oci-image/src/state.rs]: As the state is global, it depends on the `data_dir` supplied in the `ImageManager::new` arguments. The file name is a constant (`state.json`). See the `ImageManager::new`, the path is created into this method.

File: oci-image/src/state.rs
Comment (thomasgouveia) [reply oci-image/src/state.rs]: I'm not really sure to understand this point. Instead of carrying a HashMap, the `State` should carry a `snapshot_index` which will be incremented each time a mount is done ?

**Comment by thomasgouveia** [comment]
Hello @sameo,

I'm almost good with the PR. Once everything will be fixed (there are unresolved questions that I asked you in the comments above), I will re-organize my commits as you want. But I don't know how to do this separation properly, do you have a tip?

**Comment by sameo** [comment]
> But I don't know how to do this separation properly, do you have a tip?

You're going to do an interactive rebase for that.
The commit I'm asking you to split is f965406a16afd094d1904bb4ddadca12d9cf6314 and the one before it is 07b53d3417b4b95e4cb46f2462b41c045b65a0d1.

1. Start the interactive rebase, from a commit *before* the one you want to split: `git rebase -i 07b53d3417b4b95e4cb46f2462b41c045b65a0d1`
2. Here you will have git popping up an editor for you, showing something like this (with vim for me):

```
pick f965406 kaps: add crate to manage OCI images (#12)
pick 297203b kaps: add `pull` command support (#12)
pick 7a9ef51 kaps: add core logger
pick b3e8c9e kaps: add mount command support (#12)
pick c12e4c4 docs: add quickstart in README
```

You want to edit commit f965406 , so modify the above buffer to make it look like that:

```
edit f965406 kaps: add crate to manage OCI images (#12)
pick 297203b kaps: add `pull` command support (#12)
pick 7a9ef51 kaps: add core logger
pick b3e8c9e kaps: add mount command support (#12)
pick c12e4c4 docs: add quickstart in README
```

Save the buffer and let git continue, it now stops at the commit you want to modify. That should look like that:

```bash
$ git rebase -i 07b53d3417b4b95e4cb46f2462b41c045b65a0d1
Stopped at f965406...  kaps: add crate to manage OCI images (#12)
You can amend the commit now, with

  git commit --amend

Once you are satisfied with your changes, run

  git rebase --continue
```

3. You're now back in git history to the commit you want to modify. You need to unstage the changes of that commit, in order to split them into multiple ones. You unstage with `git reset HEAD^`
4. Now `git status`shows that all files from the commit you want to modify are unstaged:
```
$ git status
interactive rebase in progress; onto 07b53d3
Last command done (1 command done):
   edit f965406 kaps: add crate to manage OCI images (#12)
Next commands to do (4 remaining commands):
   pick 297203b kaps: add `pull` command support (#12)
   pick 7a9ef51 kaps: add core logger
  (use "git rebase --edit-todo" to view and edit)
You are currently splitting a commit while rebasing branch 'thomas-main' on '07b53d3'.
  (Once your working directory is clean, run "git rebase --continue")

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   .gitignore
	modified:   Cargo.lock
	modified:   Cargo.toml

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	.rustfmt.toml
	oci-image/

no changes added to commit (use "git add" and/or "git commit -a")
```

5. Now you can redo this commit into smaller ones, just as you would do regular git operations.
6. Once you're done splitting this commit into smaller ones, you can re-apply all the following commits back in by telling git to finish the interactive rebase with `git rebase --continue`
7. You're done.

Reference: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

File: oci-image/src/state.rs
Comment (sameo) [reply oci-image/src/state.rs]: `snapshot_index()`would be a method that gives you the next available index. Whenever a some asks for a mount of image docker.io/library/ubuntu:latest, you get a new index and you use it to create your mount path (e.g . `/var/run/kaps/snapshots/<index>/`). That's the path where you're snapshotter will mount the container image into.

Does that make more sense?

File: oci-image/src/state.rs
Comment (sameo) [reply oci-image/src/state.rs]: Right. I think it makes sense to give `ImageManager` the path where we want to store its metadata. But that information is specific to `ImageManager`, not to `State`. So `state.save()`should take the path as an argument, not carry it through its structure. Having it as part of the State structure is adding a field that is actually not part of the image manager state. That's why you have to `#[serde(skip)]`it. In general, when you serialize a structure and you have to explicitly skip some fields, it usually means data is not properly separated.
That's the case here: `state_file`is data about how the state itself should be managed, but its not state data.

**Comment by sameo** [comment]
> Once everything will be fixed (there are unresolved questions that I asked you in the comments above), I will re-organize my commits as you want.

I think I replied to all the pending questions. Let me know if I have not.

**Comment by thomasgouveia** [comment]
Hello @sameo, thanks for the commit splitting tip. It's exactly what I thought to do. 

Normally, I've resolved all the comments you'd addressed me. I'm missing the commit splitting. I expect to end the PR tomorrow or Sunday maximum.

Thank you for the answers!

**Comment by thomasgouveia** [comment]
Hello @sameo, everything should be ok right now. Let me know if not :)

File: oci-image/src/lib.rs
Comment (sameo) [inline oci-image/src/lib.rs]: Your constructor should not take a `snapshot_type`but a `S: dyn Snaphotter`.

It is up to the ImageManager caller and creator to decide which snapshotter it wants to use, by creating one and passing it to `ImageManager::new`.

File: oci-image/src/snapshots/mod.rs
Comment (sameo) [inline oci-image/src/snapshots/mod.rs]: `work_dir`here is:

1. Unused
2. Overlayfs specific.

I think you can remove it.

File: oci-image/src/snapshots/overlay.rs
Comment (sameo) [inline oci-image/src/snapshots/overlay.rs]: `mount`should also take a `read_only`boolean.

File: oci-image/src/state.rs
Comment (sameo) [inline oci-image/src/state.rs]: Why do you need to make this index specific to a snaphotter? Why not simply having a `sandbox_index: usize`that basically increments for every mount that is being done through the `ImageManager`, regardless of the snapshotter used?

`ImageManager`gets the next snapshot_index from its state, and passes it to the mount method of the snapshotter. Everytime state is being asked for a snapshot index, it gives the current one and increments it to fulfill the next request.

File: oci-image/src/state.rs
Comment (sameo) [inline oci-image/src/state.rs]: You really don't need that one, the only thing you care about is the next snapshot index you're going to use, regardless of the snapshotter that is being used.

File: oci-image/src/state.rs
Comment (sameo) [inline oci-image/src/state.rs]: Same here, you dont need that one.

File: oci-image/src/state.rs
Comment (sameo) [inline oci-image/src/state.rs]: That's the only one you care about. The implementation is correct, but after increasing the index, you should store it back to the state file.

File: oci-image/src/snapshots/mod.rs
Comment (sameo) [inline oci-image/src/snapshots/mod.rs]: As commented on the ImageManager and State commits, you should pass the snapshot_index to the mount method.

**Review by sameo — COMMENTED** [review]
Looking better, I still have a few comments to be addressed.

Moreover, the ordering of the commits is wrong. Actually, it is inverted. Your modules dependencies goes this way:

ImageManager depends on (State, Snapshot, Pull)
Pull depends on State
State depends on Snapshot
Snapshot depends on nothing

The way you want to order your commits is such that for every one of them you should be able to enable and use it from e.g. your main code and it should build. In other words you dont want to add a commit that depends on code that has not been committed yet.

As a consequence, your commits should be ordered in that way:

snapshot then state then pull then ImageManager. You committed them in exactly the reverse order.

**Comment by thomasgouveia** [comment]
Hi @sameo, the latest comments are all resolved. Thank you for the detailed explanation about the snapshot index, I had a misunderstanding, for me, every snapshot type had to have its own index, so it's why I chose to do a Hashmap at the start :)

File: src/cli/mount.rs
Comment (sameo) [inline src/cli/mount.rs:14]: That's not a mount command. To be addressed in a follow up PR.

Comment (sameo) [inline src/cli/mount.rs:14]: That's not a mount command. To be addressed in a follow up PR.

**Review by sameo — APPROVED** [review]

**Review by sameo — APPROVED** [review]
LGTM, nice work @thomasgouveia !

**Pull request merged**
